apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs

```diff
@@ -13,36 +13,32 @@
 // limitations under the License.
 
 using Google.Cloud.Diagnostics.Common;
-using Microsoft.AspNetCore.Http;
+using System.Threading;
 
 namespace Google.Cloud.Diagnostics.AspNetCore
 {
     /// <summary>
     /// Manages access to the current <see cref="IManagedTracer"/>.
     /// </summary>
-    internal static class ContextTracerManager
+    public static class ContextTracerManager
     {
-        ///<summary>The key to save the tracer under in the <see cref="HttpContext"/>.</summary> 
-        internal const string TraceKey = "Google.Cloud.Diagnostics.AspNetCore.Trace";
+        private static readonly AsyncLocal<IManagedTracer> _currentTracer = new AsyncLocal<IManagedTracer>();
 
         /// <summary>
         /// Sets the current <see cref="IManagedTracer"/>.  This is called in <see cref="CloudTraceMiddleware"/>
         /// for each new request.
         /// </summary>
-        public static void SetCurrentTracer(IHttpContextAccessor accessor, IManagedTracer tracer)
+        public static void SetCurrentTracer(IManagedTracer tracer)
```

Comment:
Can we leave this as internal unless you see a good reason to make it public? If we do leave it as public let's add some comments here about what the side effects will be.

Reply:
To use the tracer from other code, so code that's not in the HTTP request pipeline (e.g. a pub/sub pull handler) this will need to be public.

At Vente-Exclusive we use the class below to create a custom trace context, and as you can see it needs access to `ContextTracerManager.SetCurrentTracer`.  Perhaps it makes sense to move the `ContextTracerManager` to `Google.Cloud.Diagnostics.Common`, together with the class below -- and then make the `CloudTraceMiddleware` use them. That way the same system would be used for any trace context, in HTTP context or not.  And in that case `ContextTracerManager` could be internal again.

```csharp
public class CloudTraceContext : IDisposable
{
    private readonly ISpan _span;
    private readonly IManagedTracer _tracer;
    private readonly Action<ISpan> _annotateSpan;
    private readonly IMicroserviceContext _microserviceContext;
 
    public CloudTraceContext(string name, Action<ISpan> annotateSpan, bool? shouldTrace, IMicroserviceContext microserviceContext, Func<TraceHeaderContext, IManagedTracer> tracerFactory)
    {
        _annotateSpan = annotateSpan;
        _microserviceContext = microserviceContext;
 
        _tracer = tracerFactory(TraceHeaderContext.Create(null, 0, shouldTrace));
        ContextTracerManager.SetCurrentTracer(_tracer);
 
        if (_tracer.GetCurrentTraceId() != null)
        {
            _span = _tracer.StartSpan(name);
        }
    }
 
    public void SetStackTrace(Exception ex)
    {
        if (_tracer.GetCurrentTraceId() != null)
        {
            try
            {
                _tracer.SetStackTrace(new StackTrace(ex, true));
            }
            catch (Exception innerException)
            {
                throw new AggregateException(innerException, ex);
            }
        }
    }
 
    public void Dispose()
    {
        if (_span != null)
        {
            _annotateSpan?.Invoke(_span);

            var labels = TraceLabels.GetAgentLabel();
            labels.Add("/quantum/module", _microserviceContext.Module);
            labels.Add("/quantum/component", _microserviceContext.Component);
            labels.Add("/quantum/version", _microserviceContext.Version);
            _span.AnnotateSpan(labels);
 
            _span.Dispose();
        }
        ContextTracerManager.SetCurrentTracer(null);
    }
}
```

Reply:
For completeness, this is an example of how to use the context above:

```csharp
public async Task RunInContext(string name, Func<Task> action, Action<ISpan> annotateSpan = null, bool? shouldTrace = null)
{
    using (var context = CreateContext(name, annotateSpan, shouldTrace))
    {
        try
        {
            await action();
        }
        catch (Exception ex)
        {
            context.SetStackTrace(ex);
            throw;
        }
    }
}
```

Reply:
So what I'm trying to say is that I could submit a PR with all or parts of this together, as you think makes most sense. I'm flexible 😁

Reply:
Thanks for the extra context.  I think moving `ContextTracerManager` to common sounds like a good move.

I think I see what you mean here.  Basically a generalize version of CloudTraceMiddleware that is disposable (which ends the span)?  If so I like that.  It would open up the usage of this library.

Reply:
OK great 😃 I'll update the pull request then taking along the feedback and discussion here.

Any naming suggestions? Does `CloudTraceContext` as shown above sound good to you, or would you like to propose a different name?